### PR TITLE
Import jupyter from science overlay/ipython version bump

### PR DESCRIPTION
**Pull request by marbre**
I have written a bunch of ebuilds to bring jupyter to the science overlay. These were improved through commits/pull requests by other users/developers. Finally, they should be "stable" enough to make into the tree :)

@jlec I only send the latest versions, excluding live ebuilds. I hopefully managed it to include all required packages.

### CODE REVIEW

File: dev-python/ipython/ipython-4.0.1.ebuild [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+	use doc && emake -C docs html_noapi
+}
+
+python_test() {
+	distutils_install_for_testing
+	# https://github.com/ipython/ipython/issues/8639
+	# Failure of some modules only in python3.4 
+	local fail
+	run_tests() {
+		pushd ${TEST_DIR} > /dev/null
```
Comment (jlec) [inline dev-python/ipython/ipython-4.0.1.ebuild]: missing die

File: dev-python/ipython/ipython-4.0.1.ebuild [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+		"${PYTHON}" -m IPython.testing.iptestcontroller --all || fail=1
+		popd > /dev/null
```
Comment (jlec) [inline dev-python/ipython/ipython-4.0.1.ebuild]: missing die

File: dev-python/ipython/ipython-4.0.1.ebuild [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	VIRTUALX_COMMAND=run_tests virtualmake
+		[[ ${fail} ]] && die "Tests fail with ${EPYTHON}"
```
Comment (jlec) [inline dev-python/ipython/ipython-4.0.1.ebuild]: reduce indent

File: dev-python/ipython/ipython-4.0.1.ebuild [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+python_install() {
+	distutils-r1_python_install
+
+	# Create ipythonX.Y symlinks.
+	# TODO:
+	# 1. do we want them for pypy? No.  pypy has no numpy
+	# 2. handle it in the eclass instead (use _python_ln_rel).
+	# With pypy not an option the dosym becomes unconditional
+	dosym ../lib/python-exec/${EPYTHON}/ipython \
+		/usr/bin/ipython${EPYTHON#python}
```
Comment (jlec) [inline dev-python/ipython/ipython-4.0.1.ebuild]: @mgorny Your thoughts on this symlink, please. Is there a better way?

File: dev-python/jupyter_client/jupyter_client-4.1.1.ebuild [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+PYTHON_COMPAT=( python2_7 python3_{3,4,5} )
+PYTHON_REQ_USE="threads(+)"
+
+inherit distutils-r1
+
+DESCRIPTION="Jupyter protocol implementation and client libraries"
+HOMEPAGE="http://jupyter.org"
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+KEYWORDS="~amd64 ~x86"
```
Comment (jlec) [inline dev-python/jupyter_client/jupyter_client-4.1.1.ebuild]: Please move below line 18

File: dev-python/jupyter_core/jupyter_core-4.0.6.ebuild [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+EAPI=5
+
+PYTHON_COMPAT=( python2_7 python3_{3,4,5} )
+
+inherit distutils-r1
+
+DESCRIPTION="Core common functionality of Jupyter projects"
+HOMEPAGE="http://jupyter.org"
+SRC_URI="mirror://pypi/${PN:0:1}/${PN}/${P}.tar.gz"
+KEYWORDS="~amd64 ~x86"
```
Comment (jlec) [inline dev-python/jupyter_core/jupyter_core-4.0.6.ebuild]: As above.

File: dev-python/jupyter_core/jupyter_core-4.0.6.ebuild [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+IUSE="doc test"
+
+RDEPEND="
+	dev-python/traitlets[${PYTHON_USEDEP}]
+	"
+DEPEND="${RDEPEND}
+	doc? ( dev-python/sphinx[${PYTHON_USEDEP}] )
+	test? (
+		dev-python/pytest[${PYTHON_USEDEP}]
+		dev-python/pytest-cov[${PYTHON_USEDEP}]
```
Comment (jlec) [inline dev-python/jupyter_core/jupyter_core-4.0.6.ebuild]: Is cov really needed? It should work without.

File: dev-python/testpath/testpath-0.2.ebuild [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+EAPI=5
+
+PYTHON_COMPAT=( python2_7 python3_{3,4,5} )
+
+inherit distutils-r1
+
+DESCRIPTION="Test utilities for code working with files and commands"
+HOMEPAGE="http://jupyter.org"
+SRC_URI="https://github.com/jupyter/${PN}/archive/${PV}.tar.gz -> ${P}.tar.gz"
+KEYWORDS="~amd64 ~x86"
```
Comment (jlec) [inline dev-python/testpath/testpath-0.2.ebuild]: as above.

File: dev-python/testpath/testpath-0.2.ebuild [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	"${FILESDIR}/${P}"-setup.py.patch
+	)
+
+python_prepare_all() {
+	# Prevent un-needed download during build
+	if use doc; then
+		sed -e "/^    'sphinx.ext.intersphinx',/d" -i doc/conf.py || die
+	fi
+
+distutils-r1_python_prepare_all
```
Comment (jlec) [inline dev-python/testpath/testpath-0.2.ebuild]: Seems to be broken here.

File: dev-python/ipython/ipython-4.0.1.ebuild (hunk 4 above)
Comment (mgorny) [inline dev-python/ipython/ipython-4.0.1.ebuild]: I added it. There's no better way i know of. The alternative is for upstream to create/install symlinks/executables named like this in scriptdir.

Comment (mgorny) [inline dev-python/ipython/ipython-4.0.1.ebuild]: As a note, judging from the language of the remaining comments, idella4 changed it randomly afterwards though.

File: dev-python/ipython/ipython-4.0.1.ebuild (hunk 3 above)
Comment (marbre) [inline dev-python/ipython/ipython-4.0.1.ebuild]: @jlec All done. I also fixed this in ipython-3.2.1-r1.

File: dev-python/jupyter_client/jupyter_client-4.1.1.ebuild (hunk 5 above)
Comment (marbre) [inline dev-python/jupyter_client/jupyter_client-4.1.1.ebuild]: Done.

File: dev-python/jupyter_core/jupyter_core-4.0.6.ebuild (hunk 6 above)
Comment (marbre) [inline dev-python/jupyter_core/jupyter_core-4.0.6.ebuild]: Done.

File: dev-python/jupyter_core/jupyter_core-4.0.6.ebuild (hunk 7 above)
Comment (marbre) [inline dev-python/jupyter_core/jupyter_core-4.0.6.ebuild]: @jlec You're right. It isn't needed, I just checked upstreams travis configuration and removed it from the ebuild.

File: dev-python/testpath/testpath-0.2.ebuild (hunk 9 above)
Comment (marbre) [inline dev-python/testpath/testpath-0.2.ebuild]: @jlec I successfully executed `ebuild testpath-0.2.ebuild test` on a my laptop, on which testpath wasn't installed before. Furthermore, I am able to successfully run `ebuild ipython-4.0.1.ebuild test`, which fails if testpath isn't available.

File: dev-python/testpath/testpath-0.2.ebuild (hunk 8 above)
Comment (marbre) [inline dev-python/testpath/testpath-0.2.ebuild]: Done.

File: dev-python/testpath/testpath-0.2.ebuild (hunk 9 above)
Comment (jlec) [inline dev-python/testpath/testpath-0.2.ebuild]: ah, I miss read the code. Indention missing.

Comment (marbre) [inline dev-python/testpath/testpath-0.2.ebuild]: @jlec Fixed now :)

**Comment by jlec** [comment]
Thanks for all your work and now get your quizzes done!

I will merge this later.

**Comment by jlec** [comment]
@marbre please clean the science overlay now.

**Comment by marbre** [comment]
@jlec Cleaned the overlay. I kept the live ebuilds and also notebook-4.0.6 and ipywidgets-4.1.1, as casperjs based tests are additionally included in these.

**Pull request merged**